typify-impl/src/test_util.rs

```diff
@@ -89,6 +92,57 @@ fn validate_output_impl<T: JsonSchema + Schema>(ignore_variant_names: bool) {
     }
 }
 
+/// Reverse the canonicalization of Rust standard types performed in the codegen logic so that it
+/// is reasonable to expect that the generated Rust binding matches exactly the Rust type from
+/// which the JSON Schema was generated.
+///
+/// The code generation logic that generates Rust bindings given JSON Schema always canonicalizes
+/// any standard type (eg, `Option` is output as `::std::option::Option`), to avoid potential
+/// conflicts with types in the JSON schema with conflicting names like `Option`.  Unfortunately,
+/// this complicates the test cases that start with a Rust type that implements `JsonSchema`, use
+/// that to generate a JSON schema for that type, then use typify to generate a Rust binding for
+/// that type, and expects that the Rust AST for the source type and the generated type are exactly
+/// the same.
+///
+/// To work around this, this somewhat inelegant function simply finds every type in the input AST
+/// that starts with `::std`, and strips everything but the simple type name.  This will break if
+/// any of the input Rust types ever use fully-qualified type names that start with `::std`, but
+/// the solution is to simply not do that :)
+fn decanonicalize_std_types(mut input: DeriveInput) -> DeriveInput {
+    struct Visitor;
+
+    impl syn::visit_mut::VisitMut for Visitor {
+        fn visit_path_mut(&mut self, path: &mut syn::Path) {
+            // Check if path starts with ::std
+            if path.leading_colon.is_some()
+                && !path.segments.is_empty()
+                && path.segments[0].ident == "std"
+            {
+                // Fun additional hack to keep you on your toes:
```

Comment:
Can you check the history on that? It might have been during a previous round of prefixing

Reply:
Yes you are right, it was done in #647.

The tests that have the fully-qualified type paths in them don't seem wrong or irrelevant.  It's not a very common way to declare Rust types outside of proc macros and other codegen, but it's perfectly valid.  

I dislike the fact that a special-case is needed, and if anyone ever decides to extend those tests with a field with a fully-qualified type other than `HashMap<String, String>` then the tests will fail in a way that's probably going to be surprising.

On the other hand, it's actually not a realistic invariant that Rust -> JSON RPC -> Rust transform must always exactly reproduce the input Rust AST.  Codegen needs to fully-qualify types that would not normally be fully-qualified.  So another solution might be to run the decanonicalization pass on *both* `actual` and `expected`.  Now that I think of it, that seems like the better approach.  What do you think?

Reply:
I like the idea of running it on both actual and expected

Reply:
I just pushed https://github.com/oxidecomputer/typify/pull/705/commits/def217340862ddf63b4c1a3bbf00fc8b9af2f055 which does just that.  It's definitely better that way.